Adds Orangecrab starter project

enathang opened the pull request:
Adds orangecrab starter project based on Felix's original orangecrab project.

One choice I'm not 100% on yet is the choice to include both `Blink` and `accum` as functionality in `src/Blink.hs`. I currently include both so that users can both `make upload` to see a change in their orangecrab (blink) and also have Hedgehog tests pre-setup for them (accum). However, this might confuse the user and blink and accum are separate functions that have no relation to one another. Thoughts welcome.

enathang committed:
Adds orangecrab starter project

enathang committed:
Updates README and fixes nix flake

enathang committed:
Updates .gitignore

DigitalBrains1 commented:
> One choice I'm not 100% on yet is the choice to include both `Blink` and `accum` as functionality in `src/Blink.hs`. I currently include both so that users can both `make upload` to see a change in their orangecrab (blink) and also have Hedgehog tests pre-setup for them (accum). However, this might confuse the user and blink and accum are separate functions that have no relation to one another. Thoughts welcome.

You could do it similarly to how Christiaan did it with the DECA starter: [Tests.DECA](https://github.com/clash-lang/stack-templates/blob/6718cc240180a81efa019a0424145a5041ce3195/projects/deca/tests/Tests/DECA.hs)

DigitalBrains1 commented on projects/orangecrab/src/Orangecrab/Domain.hs:
```diff
@@ -0,0 +1,23 @@
+import Clash.Prelude
+
+-- | 48 MHz oscillator clock of the OrangeCrab board.
+createDomain vSystem
```
I think synchronous resets are usually used on Lattice ECP5?

enathang replied:
Honestly, I have no idea. So I'll look into it :)

DigitalBrains1 reviewed (commented):
Thanks for adding another nice target!

This is quite a different setup than the other starter projects. I'd rather have more uniformity; I feel it should either be similar to `simple` or to `simple-nix`. If you feel those are inadequate, let's upgrade them all. I think it's kinda hard to document or support the starter projects if they each use a different project setup and even enable different extensions in the cabal file. It's also confusing to people when they switch dev boards.

enathang committed:
Migrates clash away from clash-proxy to symlink to simple-project

enathang committed:
Bugfix: removes unused clash-proxy from cabal.project

enathang committed:
Simplifies blink example for readability

enathang committed:
Inits new project for orangecrab-nix

enathang committed:
Update proj name from orangecrab to {{name}} for stack

enathang committed:
Makes orangecrab template use {{name}} everywhere instead of hardcoded value

enathang committed:
Removes haskell from nix flake

enathang committed:
Uppercase Makefile name

enathang committed:
Adds initial draft of blink unit test

enathang committed:
Removes orangecrab-nix

enathang committed:
Changes orangecrab blink test

enathang committed:
Bugfix for blink test and removes accum

enathang committed:
Adds tests automatically ; bug fix orangecrab blink unit test

enathang committed:
Readability improvements for blink

enathang committed:
Improves orangecrab button placement and timing

DigitalBrains1 commented on projects/orangecrab-nix/src/Orangecrab/Domain.hs:
```diff
@@ -0,0 +1,23 @@
+{-|
+Module      : Domain
+Copyright   : Copyright © 2024 QBayLogic B.V.
+License     : MIT
```
Please see [here](https://github.com/clash-lang/stack-templates/blob/6718cc240180a81efa019a0424145a5041ce3195/README.md#license):
> The default license for each of the starter project is BSD2. However, this whole repository -including every starter project- is licensed under CC0. That means the authors, to the extent possible under law, have waived all copyright and related or neighboring rights to this "Clash Example Project". Feel free to choose any license for the starter projects that you want.

Would the authors of this Orangecrab code and accompanying files mind releasing it under CC0 1.0 Universal? Otherwise I think we have a bit of a problem regarding the licence. Once these files are relicensed under CC0, I'd like to suggest removing the whole licence header from all files, similar to how the other starter projects do it.

DigitalBrains1 commented on projects/orangecrab-nix/cabal.project:
```diff
@@ -0,0 +1,5 @@
+  orangecrab.cabal
+
+write-ghc-environment-files: always
+
```
```suggestion

-- Eliminates the need for `--enable-tests`, which is needed for HLS.
tests: true
```
(I just copied this from `simple-nix`. If you think it's incorrect, please correct `simple-nix`)

DigitalBrains1 commented on projects/orangecrab-nix/flake.nix:
```diff
@@ -0,0 +1,56 @@
+            serialport = dontCheck (prev.callCabal2nix "serialport" serialportSrc { });
+            hint = doJailbreak prev.hint;
+            string-interpolate = doJailbreak prev.string-interpolate;
+            orangecrab = final.callCabal2nix "orangecrab" ./. { };
```
```suggestion
            orangecrab = final.callCabal2nix "{{name}}" ./. { };
```

DigitalBrains1 commented on projects/orangecrab-nix/flake.nix:
```diff
@@ -0,0 +1,56 @@
+      {
+        devShells.default = myHsPkgs.shellFor {
+          name = "GHC 9.10.1";
+          packages = p: [ p.orangecrab ];
```
```suggestion
          packages = p: [ p.{{name}} ];
```

DigitalBrains1 commented on projects/orangecrab-nix/flake.nix:
```diff
@@ -0,0 +1,56 @@
+            ++ [ecpprog.defaultPackage.${system}]
+          ;
+        };
+        packages.default = dontCheck myHsPkgs.orangecrab;
```
```suggestion
        packages.default = dontCheck myHsPkgs.{{name}};
```

DigitalBrains1 commented on projects/orangecrab-nix/README.md:
I notice the README doesn't mention how to get a starter project. The other starter projects do contain instructions like _Run `nix-shell --packages stack --run "stack new my-clash-project clash-lang/simple-nix"`_. In fact, this README seems to not mention Nix at all.

DigitalBrains1 commented on projects/orangecrab/Makefile:
Again, I'd like to avoid `-f` if possible.

DigitalBrains1 commented on projects/orangecrab/.gitignore:
```diff
@@ -0,0 +1,8 @@
+_build/
+dist-newstyle/
+.ghc.environment.*
+*.local
```
```suggestion
/_build/
/dist-newstyle/
.ghc.environment.*
*.local
```

When you add Stack support, you should probably extend this file.

I'm looking at the difference between this file and the one for the `simple` project, and I wonder about a bunch of lines in that `.gitignore`. A bunch are clearly caught by `/_build/` but there are some others where I wonder whether those are still good to include in `simple` at all.

DigitalBrains1 commented on projects/orangecrab/README.md:
```diff
@@ -0,0 +1,48 @@
+# OrangeCrab Development Setup
+
+This sub-project contains some demo applications for targeting the [OrangeCrab FPGA r0.2.1](https://orangecrab-fpga.github.io/orangecrab-hardware/docs/r0.2.1) in combination with the OrangeCrab Pmod / Debug Expander board and an [Adafruit FT232H JTAG Debugger](https://learn.adafruit.com/adafruit-ft232h-breakout).
```
I suspect the notion of a _sub-project_ refers to where the code originates. It doesn't seem apt here.

DigitalBrains1 commented on projects/orangecrab/README.md:
```diff
@@ -0,0 +1,48 @@
+
+Some references on where to get all of the required external tooling are also given as part of the configuration file.
+
+The particular top entity to be built can be selected via overriding the `NAME` parameter in the `build.cfg.local`. Just set it to the name of the file in the `src` directory shall be built (without the `.hs` ending). See the `makefile` for the pre-configured default.
```
```suggestion
The particular top entity to be built can be selected via overriding the `NAME` parameter in the `build.cfg.local`. Just set it to the name of the file in the `src` directory shall be built (without the `.hs` ending). See the `Makefile` for the pre-configured default.
```

DigitalBrains1 commented on projects/orangecrab/README.md:
```diff
@@ -0,0 +1,48 @@
+
+#### Usage
+
+The `makefile` offers the follwing commands (to be invoked via `make CMD`) where each command depends on the output generated by all of the former ones according to the order of the list below. The `bitstream` command is the `make` default.
```
```suggestion
The `Makefile` offers the follwing commands (to be invoked via `make TARGET`) where each tareget depends on the output generated by all of the former ones according to the order of the list below. The `bitstream` target is the `make` default.
```

DigitalBrains1 commented on projects/orangecrab/README.md:
```diff
@@ -0,0 +1,48 @@
+
+The `makefile` offers the follwing commands (to be invoked via `make CMD`) where each command depends on the output generated by all of the former ones according to the order of the list below. The `bitstream` command is the `make` default.
+
+| CMD         | Description                                         |
```
```suggestion
| TARGET      | Description                                         |
```

DigitalBrains1 commented on projects/orangecrab/{{name}}.cabal:
This file differs wildly from the one in `simple`. Please use the version of `simple` and only deviate where warranted.

DigitalBrains1 reviewed (commented):
You forgot to run `render.hs`.

Could you add Stack support?

I still have more feedback, but there's no more week left now. I'll continue later; this is already a lot!

DigitalBrains1 reviewed (changes requested):
Ah woops I'm requesting changes with my review immediately above but forgot to select that again after my choice got reset through a little oopsie that cost me tons of time...

DigitalBrains1 commented:
Thanks a lot for all your work on the nice starter project! It has a well-written instructive README, it is easy to use and it adds support for a very affordable dev board. You could say a starter dev board.

Note that all my little style suggestions to the README could quickly be applied with the "Add suggestion to batch" button and then the batch can be committed. However, GitHub has a new view for the _Files_ tab and it doesn't seem to support that. If you want to be able to add suggestions to a batch, go to the _Files_ tab and in the top right corner find the "Preview - Switch back - Feedback" text. Click "Switch back", and you'll get back to the old view that has "Add suggestion to batch". Of course if instead of "Preview - Switch back" it says "✨ Try the new experience", then you're already in the old view.

I wrote:
> Could you add Stack support?

And I thought it'd be a matter of copying the `stack.yaml` from `simple` and add a few  entries to the `.gitignore`. But that's not correct! We still have the `Makefile` then which doesn't actually _use_ Stack. There's however two reasons I'd like to have Stack in there. First of all, currently, if you do `stack new my-clash-project clash-lang/orangecrab`, you get a whole bunch of warnings that are confusing to the user. Secondly, our recommended way of using Clash is with a starter project and Stack. So I'd like our actually recommended setup to work. Don't bother updating Clash 1.8.2 in `stack.yaml` to 1.8.3. Somehow, we forgot to do that when we released 1.8.3 even though it's one of the points in the checklist for releases. But we'll have a 1.8.4 Real Soon Now™ so let's tackle it then.

DigitalBrains1 commented on projects/orangecrab/build.cfg:
```diff
@@ -0,0 +1,38 @@
+CABAL=cabal
+
+# https://github.com/clash-lang/clash-compiler
+CLASH=clash
```
This var is unused.

I just noticed this; and I'm afraid I'll forget if I don't note it now. This is not yet the remainder of my review :-).

DigitalBrains1 commented on projects/orangecrab/Makefile:
This both assumes that the top entity is named literally `topEntity` and that there are no subentities with a `SYNTHESIZE` annotation (they would not be copied here). I think the README should mention these requirements.

DigitalBrains1 commented on projects/orangecrab/Makefile:
This invokes
```
cabal run clash -- src/Blink.hs
```
but that is not truly the correct thing for a project with the structure of our starter projects. You should just invoke
```
cabal run clash -- Blink
```
which will pick up the entity from the package built by Cabal.

Passing the source file means you again compile `Blink.hs` from source and only its imports come from pre-compiled stuff. This also means none of the things in `orangecrab.cabal` or `cabal.project` apply to `Blink.hs`, it is compiled with the builtin settings of the Clash compiler. There are situations where the two are simply not equivalent at all.

DigitalBrains1 commented on projects/orangecrab/src/Orangecrab/Domain.hs:
```diff
@@ -0,0 +1,23 @@
+{-|
+Module      : Domain
+Copyright   : Copyright © 2024 QBayLogic B.V.
+License     : MIT
```
> Licenses apply as given by these repositories.

Are you saying you don't want to relicense your makefile to be compatible with this repository's license? If so, how do you think we should proceed?

DigitalBrains1 replied:
I'd say this part of the MIT licence

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

prevents recipients of the licensed work from redistributing it under CC0 (or potentially other licences; the whole idea is that the user of the starter project is unencumbered in their choice of licence for their project).

But it's fairly inconsequential here; if you, the author, are okay with releasing the Makefile under CC0, that is all that needs to be established here in this PR.

Thank you!

DigitalBrains1 commented:
Ah, yes, you should definitely not consider that a released `clash-cores`.

enathang committed:
Update comments from Peter feedback

enathang committed:
Updates Makefile with Peter's proposed change

enathang committed:
adds orangecrab.hsfiles

DigitalBrains1 committed:
Remove the Clash step from the Makefile

DigitalBrains1 committed:
Renumber build dirs

Since we dropped a step, it started at 02-hdl. Renumber to start at
01-hdl and on.

DigitalBrains1 committed:
Adopt .cabal file from simple project

And make the tests have the same layout as there

DigitalBrains1 committed:
Add stack.yaml

DigitalBrains1 committed:
Remove LICENSE

We don't have one in the other starter projects, we specify 2-BSD not
MIT, and people might use any license they choose.

DigitalBrains1 committed:
Adopt/adapt simple's gitignore

Added are
/_build/
*.local

The latter came in the place of the overlapping /cabal.project.local.

DigitalBrains1 committed:
Merge in stuff from simple README

Also add some useful information and do some copyediting.

DigitalBrains1 committed:
Rerun render.hs

DigitalBrains1 committed:
Change domain to sync resets

An alternative would be to use `vXilinxSystem` instead of `vSystem` as
the basis for the record update; but I think this is more confusing than
helpful even though it's shorter.

enathang committed:
Makefile formatting fixes

DigitalBrains1 committed:
README: no Nix, no issues with init values

DigitalBrains1 committed:
Fix polarity issue in `driveRGB`

DigitalBrains1 committed:
Get rid of license headers

The whole repo is CC0.

DigitalBrains1 committed:
Remove Nix stuff

enathang committed:
[skip-ci] Re-runs ./render

enathang committed:
Fixes unit test failing

DigitalBrains1 committed:
README: add missing closing parenthesis

DigitalBrains1 committed:
Fix somewhat misleading test step

DigitalBrains1 committed:
Invoke ./render.hs

enathang committed:
Fixes ci for orangecrab

DigitalBrains1 reviewed (approved):
(You might already have noticed, but the CI scripts should match on `orangecrab.hsfiles` ending in an s, not `.hsfile`)

enathang committed:
Fix typo in .ci bash scripts that was failing orangecrab ci
